File: MultiLanguageSandbox/src/test/deps/r/30.R



# This function, minimum_score_merge, calculates the minimum score required to merge N piles of stones into a single pile. The score for each merge is the sum of the stones in the two piles being merged. The function takes a vector representing the number of stones in each pile and returns the minimum total score needed for all merges.
# Examples:
# >>> minimum_score_merge(c(13, 7, 8, 16, 21, 4, 18))
#     239

minimum_score_merge <- function(stone_piles) {



    total_score <- 0
    piles <- as.list(stone_piles)
    while (length(piles) > 1) {
        piles <- sort(unlist(piles), decreasing = FALSE)
        piles <- as.list(piles)
        first <- piles[[1]]
        second <- piles[[2]]
        piles <- piles[-c(1, 2)]
        merge_score <- first + second
        total_score <- total_score + merge_score
        piles <- c(piles, list(merge_score))
    }
    
    return(total_score)
}
# Test cases
main <- function() {

    stopifnot(minimum_score_merge(c(13, 7, 8, 16, 21, 4, 18)) == 239)
    stopifnot(minimum_score_merge(c(10, 20, 30)) == 90)
    stopifnot(minimum_score_merge(c(1, 2, 3, 4)) == 19)
}


# Run tests
main()

File: MultiLanguageSandbox/src/test/deps/ruby/30.rb


# Calculates the number of possible photo arrangements for a group of people.
# In this arrangement, people are lined up in such a way that the age difference
# between any two adjacent persons is no more than two years.
# The function takes a single integer argument, n, representing the number of people
# in the group, where each person has a distinct age from 1 to n years.
#
# Example:
# >>> photo_arrangements_count(4)
# 4
# >>> photo_arrangements_count(5)
# 6

def photo_arrangements_count(n)
  return 0 if n == 0
  return 1 if n == 1

  # We'll use memoization to store the number of ways to arrange the remaining numbers
  # with the last number being 'last' and the set of used numbers.
  # However, for n up to 10 or so, this might be feasible with memoization.
  # For larger n, a more optimized approach is needed, but given the examples,
  # perhaps n is small.

  # Initialize a memo hash to store intermediate results
  memo = {}

  def dfs(used, last, n, memo)
    key = [used.sort, last]
    return memo[key] if memo.key?(key)

    if used.size == n
      return 1
    end

    count = 0
    (1..n).each do |num|
      next if used.include?(num)
      if used.empty? || (num - last).abs <= 2
        new_used = used.dup
        new_used << num
        count += dfs(new_used, num, n, memo)
      end
    end

    memo[key] = count
    count
  end

  total = 0
  (1..n).each do |start|
    used = [start]
    total += dfs(used, start, n, memo)
  end

  total
end

  
raise 'Test failed' unless photo_arrangements_count(4) == 4
raise 'Test failed' unless photo_arrangements_count(5) == 6
raise 'Test failed' unless photo_arrangements_count(6) == 9
raise 'Test failed' unless photo_arrangements_count(7) == 14
raise 'Test failed' unless photo_arrangements_count(8) == 21
  

puts 'All tests passed!'